### ISSUE TITLE
Meteor should transpile js in node_modules

### ISSUE THREAD

**Issue by nathantreid** [meteor/meteor#8640]
The problem: due to the popularity of Webpack, many npm modules are being released that expect to be transpiled during the app build. I've run into this numerous times lately, for example a number of the `vue-bulma` packages use import/export, and just today I ran into the issue with [vue-fraction-grid](https://github.com/bkzl/vue-fraction-grid). Vue-fraction-grid provides both a compiled version in the `dist` folder and the source for use with Webpack. Unfortunately, Meteor reads all of the package's files, so it dies with the following error when processing the source: 
```
W20170426-11:49:08.176(-5)? (STDERR) export default {
W20170426-11:49:08.176(-5)? (STDERR) ^^^^^^
W20170426-11:49:08.176(-5)? (STDERR) 
W20170426-11:49:08.176(-5)? (STDERR) SyntaxError: Unexpected reserved word
```

From what I understand, there are 2 reasons why Meteor doesn't transpile node_modules:
1. Performance
2. To avoid potential issues from transpiling files that aren't intended to be transpiled.

I propose that Meteor either allow opt-in transpilation via adding a plugin (`meteor add  ecmascript-node-modules`), or preferably implement some sort of configuration that determines which modules to process either by adding a local package (`meteor-ecmascript-config`) or preferably by adding configuration options in `package.json` or a `.meteorconfig` file.
Either the package or the config file would be used to set options such as 
``` json
{
  "transpileNodeModulesJs": true
}
```
or
``` json
{
  "transpileNodeModulesJs": ["vue-fraction-grid"]
}
```

Today, a workaround for this is to manually install the files into your project, but this is quite a hassle.
By implementing this change, Meteor can maintain backwards compatibility while greatly widening the scope of packages that can be easily used.

**Comment by benjamn** [comment]
In Meteor 1.5, we're experimenting with this idea for packages that have a "module" field in their `package.json` files, though the current implementation compiles only ECMAScript module syntax (`import` and `export`, as in your example): https://github.com/meteor/meteor/pull/8327/commits/d7cae3c939e2231ed89e694d05fe5f3698540b72

You can try out that functionality by running `meteor create --release 1.5-beta.17 test-app`. Please let us know if it works!

**Comment by benjamn** [comment]
Specifically, in the latest Meteor 1.5 beta, you should now be able to import any `node_modules` that use `import` and `export` (but no other exotic ECMAScript features):
```js
import VueFractionGrid from "vue-fraction-grid/vue-fraction-grid.js"
```
However, I think you'll need to have a Meteor compiler plugin that handles `.vue` files in order for [these imports](https://github.com/bkzl/vue-fraction-grid/blob/6b0bc07665c896bc9a83fcb003f81f3e50dacb51/vue-fraction-grid.js#L1-L3) to succeed.

**Comment by nathantreid** [comment]
@benjamn I've been testing this out and I'm running into issues.
With both `vue-fraction-grid` and `vue-multiselect`, I receive `Unexpected reserved word` errors for `import` and `export` keywords.
Here's a reproduction with `vue-multiselect`: https://github.com/nathantreid/meteor-1.5-node-modules-import. `vue-multiselect` uses the `jsnext:main` attribute, which from your referenced commit I think should work just as well as `module`. I tried changing it to `module` just to see, but that had no effect.

The files that cause the error (in the `src` folder) aren't actually being used, but Meteor attempts to parse them and errors on the import/export statements. I tried importing the `src/index.js` from `dist/vue-multiselect.min.js` in case that was why Meteor was failing to apply the new import/export functionality, but that had no effect.
The only thing that has worked so far was for me to clone the package and edit the `.npmignore` file so it only includes the `dist` folder.

**Comment by eybarta** [comment]
@nathantreid I've been having the same issues with Vue npm packages, vue-multiselect and now with vue-dropzone, could you elaborate on your work around with editing the .npmignore? I couldn't make it work.

**Comment by nathantreid** [comment]
@eybarta, Here's what I did for `vue-multiselect`: https://github.com/monterail/vue-multiselect/compare/2.0...nathantreid:2.0.
Then I published a scoped version of the package (@nathantreid/vue-multiselect) so I could still `npm install` it. If you go down the same route, and you're publishing it as a public package, the command is `npm publish --access public`. You only need the `--access public` if you're publishing it as a public scoped package, and you only need it the first time you publish the package.
The downside to this, of course, is that I now have to manually update my version of the package whenever I want to get updates.

**Comment by eybarta** [comment]
@nathantreid thanks. It seems kind of a pain to republish a plugin just for build errors ;\
Somehow specifically for vue-multiselect, the errors went away after I uninstalled/reinstalled, and erased all Meteor cache.. not sure what's happening there.. 
What I did at first, when I couldnt' solve this, was to copy the Vue files as is from the npm folder and make it a local vue component inside my imports folder, that also works.

**Comment by mwarren2** [comment]
I'm unable to `npm install quasar-framework` (a vue-based ui framework).
I use akryum:vue and akryum:vue-component to hook up vue to meteor, and then I try to install quasar.
I hope very much that this issue will eventually make this possible.

```
W20170602-14:50:29.700(2)? (STDERR) /Users/malcolm/git/app-template-meteor/.meteor/local/build/programs/server/packages/modules.js:372
W20170602-14:50:29.701(2)? (STDERR) import TableFilter from './TableFilter.vue'
W20170602-14:50:29.702(2)? (STDERR) ^^^^^^
W20170602-14:50:29.702(2)? (STDERR) 
W20170602-14:50:29.713(2)? (STDERR) SyntaxError: Unexpected reserved word
```

**Comment by hwillson** [comment]
@mwarren2 I just cloned the [`Akryum/vue-meteor-demo`](https://github.com/Akryum/vue-meteor-demo) app, then ran `meteor npm install --save quasar-framework`. I was able to install the framework properly, and the demo app starts and runs properly. Might be something else causing this in your app?

**Comment by mwarren2** [comment]
Did you try to import quasar and use it?

I can npm install quasar-framework into the demo, and launch meteor ok, but nothing more.
If I try to import quasar, then it wants fastclick. If I install fastclick, then I get "ReferenceError: navigator is not defined" which I have been unable to solve.

So I have been unable even to try quasar properly in the demo.

**Comment by mwarren2** [comment]
@hwillson Would you mind trying this app instead?
https://github.com/mwarren2/try-app/

**Comment by hwillson** [comment]
To help provide a more clear separation between feature requests and bugs, and to help clean up the feature request backlog, Meteor feature requests are now being managed under the https://github.com/meteor/meteor-feature-requests repository. 

This feature request will be closed here, but anyone interested in migrating this feature request to the new repository (to make sure it stays active), can click [here](https://github.com/meteor/meteor-feature-requests/issues/new?title=Meteor%20should%20transpile%20js%20in%20node_modules&body=Migrated%20from:%20meteor/meteor%238640) to start the feature request migration process. This manual migration process is intended to help identify which of the older feature requests are still considered to be of value to the community. Thanks!

**Comment by benjamn** [comment]
We think we've finally found a way to solve this problem: https://github.com/meteor/meteor/pull/9771

**tl;dr** Use symbolic links to expose `node_modules` code within your application, outside of `node_modules`. Meteor will compile the exposed code as if it was part of your application, using whatever compiler plugins you have installed, and also guarantee that you get the compiled code when you import from `node_modules` (this is they key new feature).
